core/trino-main/src/main/java/io/trino/server/security/oauth2/JWTTokenSerializer.java

```diff
@@ -0,0 +1,158 @@
+import io.jsonwebtoken.JwtParser;
+import io.jsonwebtoken.SignatureAlgorithm;
+import io.trino.client.NodeVersion;
+
+import java.security.Key;
+import java.time.Clock;
+import java.util.Date;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.trino.server.security.jwt.JwtUtil.newJwtBuilder;
+import static io.trino.server.security.jwt.JwtUtil.newJwtParserBuilder;
+import static io.trino.server.security.oauth2.JWTTokenSerializer.ZstdCodec.CODEC_NAME;
+import static java.lang.Math.toIntExact;
+import static java.util.Arrays.copyOfRange;
+import static java.util.Objects.requireNonNull;
+
+public class JWTTokenSerializer
+        implements TokenPairSerializer
+{
+    private static final CompressionCodec COMPRESSION_CODEC = new ZstdCodec();
+    private static final String ACCESS_TOKEN_KEY = "access_token";
+    private static final String REFRESH_TOKEN_KEY = "refresh_token";
+    private final Key signingKey;
+    private final Key verificationKey;
+    private final SignatureAlgorithm signatureAlgorithm;
+    private final OAuth2Client client;
+    private final io.jsonwebtoken.Clock clock;
+    private final String issuer;
+    private final JwtParser parser;
+
+    public JWTTokenSerializer(
+            JwtKeysProvider keysProvider,
+            SignatureAlgorithm signatureAlgorithm,
+            OAuth2Client client,
+            NodeVersion nodeVersion,
+            Clock clock,
+            long expirationTimeExtensionSeconds)
+    {
+        this.signingKey = requireNonNull(keysProvider, "keysProvider is null").getSigningKey();
+        this.verificationKey = requireNonNull(keysProvider, "keysProvider is null").getVerificationKey();
+        this.signatureAlgorithm = requireNonNull(signatureAlgorithm, "signatureAlgorithm is null");
+        this.client = requireNonNull(client, "client is null");
+        this.issuer = requireNonNull(nodeVersion, "nodeVersion is null").getVersion();
+        this.clock = () -> Date.from(requireNonNull(clock, "clock is null").instant());
+
+        this.parser = newJwtParserBuilder()
+                .setSigningKey(verificationKey)
+                .setClock(this.clock)
+                .requireIssuer(this.issuer)
+                .setCompressionCodecResolver(JWTTokenSerializer::resolveCompressionCodec)
+                .setAllowedClockSkewSeconds(expirationTimeExtensionSeconds)
+                .build();
+    }
+
+    @Override
+    public TokenPair deserialize(String token)
+    {
+        requireNonNull(token, "token is null");
+
+        try {
+            Claims claims = parser.parseClaimsJws(token).getBody();
+            return TokenPair.accessAndRefreshTokens(
+                    claims.get(ACCESS_TOKEN_KEY, String.class),
+                    claims.get(REFRESH_TOKEN_KEY, String.class));
+        }
+        catch (ExpiredJwtException ex) {
+            throw new IllegalArgumentException("Token has timed out", ex);
```

Comment:
I understand that we implicitly capture when the access token expires, however I'm asking here that it be explicit instead. The current value for the token expiration is good, having it expire some time after the access token itself should expire is good. However I would also like to see the token expiration encoded into the token and used to force a refresh of the token rather than us having to find out the hard way by retrieving the claims associated with the token.